modules/apps/depot/depot-web/src/main/resources/META-INF/resources/screen/navigation/entries/sites.jsp

```diff
@@ -33,11 +33,20 @@ List<DepotEntryGroupRel> depotEntryGroupRels = depotAdminSitesDisplayContext.get
 			<span class="heading-text"><liferay-ui:message key="connected-sites" /></span>
 		</clay:content-col>
 
+		<%
+		Map<String, String> additionalPropsData = new HashMap<>();
+
+		additionalPropsData.put("currentURL", currentURL);
+		additionalPropsData.put("eventName", liferayPortletResponse.getNamespace() + "selectSite");
+		additionalPropsData.put("itemSelectorURL", String.valueOf(depotAdminSitesDisplayContext.getItemSelectorURL()));
+		%>
+
 		<clay:content-col>
 			<span class="heading-end">
 				<clay:button
+					additionalProps="js/addConnectedSiteButtonAdditionalProps"
+					additionalPropsData="<%= additionalPropsData %>"
```

Comment:
Hideous 2 param passing... thoughts?

Maybe we could provide a higher level abstraction that had `{src, data}` so we could just pass that thing. It would complicate the case where only `namespace` or default passed data is needed since it will be no longer possible to just point to a readable path... it also makes it more verbose and we will probably end up hiding the `js/*Props` path inside a java DisplayContext class...

Reply:
I think it's hideously only because of the naming. It seems ugly to have two props with names like `additionalProps` and `additionalPropsData` only because it sounds like they are two half-baked semi-overlapping things that aren't complete without each other.

But if you pick different names, the smell of 💩 starts to become a smell of 🌹 ... maybe. It could be something like `propsTransformer` (the module exporting the JS function) and `additionalProps` (or some similar name, for the props data blob).

Reply:
Mmm... so 

```jsp
<clay:button
    propsTransformer="js/buttonPropsTransformer"
    additionalProps="<%= additionalProps %>"
/>
```

I'd assume with this that if no `propsTransformer` is passed, `additionalProps` will still be passed straight to the component. And also stemming from there, I'd expect all props being passed to the component to go through the `propsTransformer` and not just the `additionalProps`... do we all agree on that?

Reply:
> I'd assume with this that if no `propsTransformer` is passed, `additionalProps` will still be passed straight to the component. And also stemming from there, I'd expect all props being passed to the component to go through the `propsTransformer` and not just the `additionalProps`... do we all agree on that?

That's what I'd expect too. [Principle of Least Surprise](https://wiki.c2.com/?PrincipleOfLeastAstonishment).

Reply:
Okay, feels like we're getting there... `propsTransformer` still doesn't exactly click... in many cases it won't transform anything but add more... like a `propsDeriver` or `propsFromDataDeriver`.

Now... one other question would be...

```jsp
// additionalProps: {fizz: 'fizz', buzz: 'buzz'}
<clay:button
    size="lg"
    displayType="unstyled"
    propsTransformer="js/buttonPropsTransformer"
    additionalProps="<%= additionalProps %>"
/>
```

```js
export default propsTranformer({size, displayType, foo, bar, namespace}) {
    return {
        fizzbuzz: `${fizz}${buzz}`
    }
```

```js
export default ({...otherProps}) => {
    // Which one? :)

    console.log(otherProps); // size, displayType, fizz, buzz, namespace, fizzbuzz
    console.log(otherProps); // size, displayType, fizzbuzz
    console.log(otherProps); // fizzbuzz
}
```

Reply:
I think I'd expect all props and the developer can do whatever they want with it.

```js
export default propsTranformer(props) {
    return {
        ...props,
        fizzbuzz: `${props.fizz}${props.buzz}`,
    }
}
```

Reply:
I was expecting `propsTransformer` to take `taglibAttributes`, `additionalProps` (whatever the names end up being) and merge them as it sees fit.

```js
export default propsTransformer(taglibAttributes, additionalProps = {}) {
   // this, or any arbitrary transformation that I want to make
   // (merging, overriding, removing, adding, modifying etc)
   return {
        ...taglibAttributes,
        ...additionalProps
   };
}
```

The naming is the hard bit, of course.

Reply:
Ah okay. So we don't necessarily merge before passing it to `propsTransformer` right?

Reply:
So, my only concern is that if someone wants to add an `onClick` then they need to do this:

```js
export default propsTransformer(taglibAttributes, additionalProps = {}) {
   return {
        ...taglibAttributes,
        onClick() { ... }
   };
}
```

It feels too error-prone because they might be focused on just adding an `onClick` prop and then forget to pass in the others... am I being over-protective here?

Reply:
I think I'd rather have this than having the merge be done automatically and taking away the freedom of manipulating it however I want.

Reply:
@wincent, @brunobasto, @bryceosterhaus, I just realized we already support this in `react:component` as a `data` attribute, so that feels just like what we want. Would we be okay with `data` instead of `additionalProps`, or should we move `react:component` to use `props` instead of `data` to keep things a bit more consistent?

Not sure about this yet, but it might make sense that we'd also want to add `propsTransformer` support to the `react:component` tag... 🤔

Reply:
> I just realized we already support this (and by this I just mean `additionalProps`) in `react:component` as a `data` attribute, so that feels just like what we want. Would we be okay with `data` instead of `additionalProps`, or should we move `react:component` to use `props` instead of `data` to keep things a bit more consistent?

I'd think we'd best not change `react:component` (officially would be breaking, or do you think we can get away with it?). I'd be fine with using `data` here (it's shorter) if you want. It doesn't bother me that the React-specific terminology ("props") ends up mapping onto something else ("data") at some point on the journey between Java/JSP-land and React-land, although in an ideal world, it would be lovely if we had top-to-bottom consistency in the naming.
 
> Not sure about this yet, but it might make sense that we'd also want to add `propsTransformer` support to the `react:component` tag... 🤔

Maybe wait and see? YAGNI and all that.

Reply:
> [...] officially would be breaking, or do you think we can get away with it? [...]

We can't and we shouldn't. What we could do would be to deprecate the `data` attribute, add a `props` one and replace usages in our code.

> Maybe wait and see? YAGNI and all that.

Yeah, definitely

Reply:
> @wincent, @brunobasto, @bryceosterhaus, I just realized we already support this (and by this I just mean additionalProps) in react:component as a data attribute, so that feels just like what we want. Would we be okay with data instead of additionalProps, or should we move react:component to use props instead of data to keep things a bit more consistent?
> 
> Not sure about this yet, but it might make sense that we'd also want to add propsTransformer support to the react:component tag... 🤔

Humm, right. What Greg said plus in Liferay the `data` attribute for TagLibs was used for a while to mean that it will add a data attribute `data-foo="bar"` to the markup. So it maybe confusing for some people.

Reply:
> Humm, right. What Greg said plus in Liferay the data attribute for TagLibs was used for a while to mean that it will add a data attribute data-foo="bar" to the markup. So it maybe confusing for some people.

That's right, thanks for the reminder @brunobasto!! 

Now that the `clay` tags are dynamic, devs can add any attribute they want to it directly, which I think makes it easier to reason with and makes it obvious that `data-foo` goes to the container and is not a prop.

I got tripped by this myself just now! 😂

Reply:
Ok, so this is half done. At least the internals are done at [Renames data as props to provide the proper semantics in this context](https://github.com/liferay-frontend/liferay-portal/pull/2/commits/5420c6caf0f275f352be2eef8b00982c19ba7bda) and I've created [Provide props API to replace data attribute in react:component tag](https://issues.liferay.com/browse/LPS-116504) to track flipping the API in `react:component`